File: GitHubPop/app/src/androidTest/java/com/igormeira/githubpop/RepositoryListTest.java
package com.igormeira.githubpop;

import android.content.Context;

import androidx.recyclerview.widget.RecyclerView;
import androidx.test.espresso.contrib.RecyclerViewActions;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.rule.ActivityTestRule;

import com.igormeira.githubpop.repository.RepositoriesActivity;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static androidx.test.espresso.Espresso.onData;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static com.igormeira.githubpop.Helper.waitFor;
import static org.hamcrest.Matchers.startsWith;
import static org.junit.Assert.*;

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(AndroidJUnit4.class)
public class RepositoryListTest {
    @Rule
    public ActivityTestRule<RepositoriesActivity> rule  = new  ActivityTestRule<>(RepositoriesActivity.class);

    @Test
    public void showingRepositories(){
        waitFor(2000);
        onView(withText(startsWith("CyC2018")))
                .check(matches(isDisplayed()));
    }
}

File: app/src/androidTest/java/dev/renatoneto/githubrepos/ui/RepositoryListTest.kt
package dev.renatoneto.githubrepos.ui

import android.content.Intent
import android.os.SystemClock
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.*
import androidx.test.espresso.assertion.ViewAssertions.*
import androidx.test.espresso.contrib.RecyclerViewActions.*
import androidx.test.espresso.intent.Intents.*
import androidx.test.espresso.intent.matcher.IntentMatchers.*
import androidx.test.espresso.intent.rule.IntentsTestRule
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.runner.AndroidJUnit4
import dev.renatoneto.githubrepos.R
import dev.renatoneto.githubrepos.base.BaseTest
import dev.renatoneto.githubrepos.ui.repositorydetails.RepositoryDetailsActivity
import dev.renatoneto.githubrepos.ui.repositorylist.RepositoryListActivity
import org.hamcrest.CoreMatchers.*
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Instrumented test, which will execute on an Android device.
 *
 * See [testing documentation](http://d.android.com/tools/testing).
 */
@RunWith(AndroidJUnit4::class)
class RepositoryListTest : BaseTest() {

    @get:Rule
    val rule = IntentsTestRule(
        RepositoryListActivity::class.java, true, false
    )

    @Test
    fun testSuccessLoadRepositories() {
        rule.launchActivity(Intent())

        enqueueMockServer(rule.activity, "repository_list.json")

        onView(withId(R.id.progressBar)).check(matches(isDisplayed()))

        SystemClock.sleep(300L)
        onView(allOf<View>(withId(R.id.txtRepositoryName), withText("java-design-patterns")))
            .check(matches(isDisplayed()))

        rule.finishActivity()
    }

    @Test
    fun testErrorLoadRepositories() {
        rule.launchActivity(Intent())

        enqueueMockServer(rule.activity, "repository_list.json", 404)

        SystemClock.sleep(300L)
        onView(withId(R.id.errorLayout)).check(matches(isDisplayed()))

        rule.finishActivity()
    }

    @Test
    fun testGoToRepositoryPullRequests() {
        rule.launchActivity(Intent())

        enqueueMockServer(rule.activity, "repository_list.json")

        SystemClock.sleep(300L)
        onView(withId(R.id.recyclerRepositories)).perform(
            actionOnItemAtPosition<RecyclerView.ViewHolder>(0, click())
        )

        intended(hasComponent(RepositoryDetailsActivity::class.java.name))

        rule.finishActivity()
    }

}
